### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by ivangusev** [boazsegev/combine_pdf#51]
I'm running this gem for my project where i need to combine multiple pdf into single pdf file.
It seems ruby process takes more and more memory with every pdf.

**Comment by boazsegev** [comment]
Hi Ivan, thank you for opening the issue.

Ruby has a garbage collector, so a memory "leak" (memory bloat, actually) means that old objects are still being referenced somewhere.

Could you provide some example code for testing? I would like to validate the issue and see where this might be happening.

**Comment by ivangusev** [comment]
Hi Bo, thanks for reply.

I have a ruby daemon process. Here is example how i use "combine_pdf" gem.
Then i review memory usage.
If i comment out "combine" code then memory usage stay on same level without growing.

```
require 'daemons'

Daemons.run_proc('myp.rb',{:dir_mode=>:script,:dir=>'logs',:log_output=>true}) do
i=1
loop do
 	require 'combine_pdf'
    first_page_pdf='1.pdf' # always refer to first pdf
    pyon_page_pdf=i.to_s+'.pdf' # refer to next pdf
    # combine code:
    (CombinePDF.load(first_page_pdf) << CombinePDF.load(pyon_page_pdf)).save(first_page_pdf)
    i+=1
end

end
```

**Comment by boazsegev** [comment]
Hi Ivan,

I'm not sure I'm getting this right... maybe because this is just the example code.

Your code is loading the same file (`1.pdf`), combining it with the second file (`"#{i}.pdf"`), and saving the whole thing back to `1.pdf`, so that `1.pdf` is always growing... but... when does the loop exit? and what happens if the `"#{i}.pdf"` file doesn't exist? does the task restart (bloating `1.pdf` even more)?

If this is always happening (if your loop never ends), every time you load the file from the disk (which I'm not sure why you're doing), you're re-parsing the file and storing it back in the memory as a new instance.

This means that now you will have a new parser instance and a new PDF instance **for every loop cycle**.

Even after the Ruby's GB (garbage collector) releases the old PDF instances (the previous `1.pdf` PDF objects and parsers that the previous loops used), you're still always adding more pages to the `1.pdf` file and it's always growing and there's at least one copy of the file always in the memory.

From what I can tell, your memory consumption is growing because your loop actually needs more memory.

I think the design should be re-evaluated.

In addition to re-evaluating the design, I would consider the following adjustments:

```ruby
require 'daemons'

Daemons.run_proc('myp.rb',{:dir_mode=>:script,:dir=>'logs',:log_output=>true}) do

# requiring the library once is enough - move this out of the loop.
require 'combine_pdf'
# you can use the same object again and again - this is both faster
# and avoids the memory consumption required when parsing new copies.
pdf_1 = CombinePDF.load('1.pdf')
i=1
loop do
    pyon_page_pdf=i.to_s+'.pdf' # refer to next pdf
    # prevent exceptions
    if (File.file? pyon_page_pdf)
       (pdf_1 << CombinePDF.load(pyon_page_pdf)).save(first_page_pdf)
    i+=1
    else
        break
    end
end
end

```

This would mean the loop exists when all the pages are done.

please notice that loading a nonexistent file will throw an exception, so the deamon might try to restart the process, bloating `1.pdf` even more with already existing copies (but it might not, this depends on the deamon's implementation).

Also, CombinePDF will use all the memory required for keeping the PDF object in the memory AND for creating a string with the formatted PDF data that is outputted to the file (This temporary string object is created whenever you call `save`)... So - even after these adjustments - I expect your memory consumption to be about x2-x3 the size of the PDF file (at least).

Please let me know if the issue persists and if you have and questions.

Good Luck!

**Comment by ivangusev** [comment]
Hi Bo, thanks for reply.
It was just example. My daemon code is very big and coupled with other "systems".
But idea is that after pdf file is saved - memory should be released, i don't understand why you keep it in memory.
My need is very simple: i take one pdf, and combine it with other 5 pdf. After it i expect memory is back to initial state, but in reality no.

**Comment by boazsegev** [comment]
Hi Ivan,

I performed the following simple test on my machine:

```ruby
require 'combine_pdf'
1000.times { CombinePDF.load(filename) }
```

If the memory was leaking, my Ruby VM should have been bloating, but I got only the normal memory increase that Ruby performs.

I believe the issue is with your code, but I may be mistaken (since all you provide me with is examples).

For example:

* Your code doesn't show an exit point. What if 4.pdf doesn't exist and CombinePDF throws an exception trying to load the file?  Does the task restart? What happens?

* Why are you reloading a file - if you need the same file, why not keep the loaded version?

* Why are you saving each time you loop?

* Why are you requiring the library again and again inside the loop?

Consider the following adjustment to the code you posted:
```ruby
require 'daemons'

Daemons.run_proc('myp.rb',{:dir_mode=>:script,:dir=>'logs',:log_output=>true}) do

# requiring the library once is enough - move this out of the loop.
require 'combine_pdf'
# you can use the same object again and again - this is both faster
# and avoids the memory consumption required when parsing new copies.
pdf_1 = CombinePDF.load('1.pdf')
i=1
loop do
    pyon_page_pdf=i.to_s+'.pdf' # refer to next pdf
    # perform CombinePDF.load ONLY if the file exists.
    if (File.file? pyon_page_pdf)
       # Add the next file to the PDF.
       (pdf_1 << CombinePDF.load(pyon_page_pdf))
       i+=1
    else
       # save only once, when we're finished.
        pdf_1.save("output_file.pdf")
        # load the new file
        pdf_1 = CombinePDF.load('new_pdf_file.pdf')
        i = 1
    end
end
end
```

Again, there is no need to load the same file in a loop. Move the loading outside the loop. Loading the file and parsing the data is more expensive then using the data you already have (you already loaded the file once, why do it again?).

---

Please notice that under normal operations, [Ruby's GC (garbage collection)](http://ruby-doc.org/core-2.3.0/GC.html) does NOT release all the memory back to the OS. The Ruby interpreter assume that the memory will be needed again (after all, your program needed the memory before) and it will release memory slowly over time, keeping some of the memory for future use.

You can read more about the Ruby GC by performing a search on line.

**Comment by ivangusev** [comment]
Hi Bo,

In your example there are "unclear" places, which are "filename" (It should not be same!), missing "save" method. Also without daemon is problematic to notice memory issue, because after ruby script is finished memory is released.
Once again, i'm trying to explain issue, i take A.pdf, combine it with B.pdf, script is sleeping(waiting), memory usage 30Mb, i take M.pdf(with size less than A.pdf), combine it with N.pdf, script is sleeping, memory usage 50MB, if i continue this process then all free memory will be taken.

The only solution i have for now is to replace your gem with "ghostscript".
But all i need is to combine pdf files, so your gem for me would be better solution.

**Comment by boazsegev** [comment]
What version of CombinePDF are you using? because I can't reproduce the issue.

You're right about my example... It was late when I wrote you back and I didn't copy correctly.

I didn't write the file name because It contains the full path and user name on my machine... but you can use any file name (you'll need two existing PDF files and a third name).

I did save every time the loop was running:
```ruby
require 'combine_pdf'
filename1 = `1.pdf`
filename2 = `2.pdf`
filename1 = `out.pdf`
1000.times { (CombinePDF.load(filename1) << CombinePDF.load(filename2)).save(filename3) }
```

I ran the example in `irb`, so the ruby process did't exit after the loop was finished.

Also, because I chose big PDF files (250Kb each), the loop took a long time and I could test the memory growth while running the command. The process, on my machine, grew to ~60Mb and stopped growing (the loop continued).

Please, feel free to test a simple loop - your code, **without the deamon** and using PDF files on my machine, doesn't cause the memory to bloat on my machine (unless I keep combining to the same file and then the PDF grows all the time, so it's using more and more memory).

**Comment by boazsegev** [comment]
Ivan,

I'm sorry if it feels like I'm not listening.

I appreciate that you opened the issue, I just can't reproduce it and I'm trying to figure out if the issue is related to CombinePDF or other things (like the deamon or the code in the loop)...

I am looking into this issue and I'm trying to locate any potential memory leaks.

I thank you for your efforts.

**Comment by boazsegev** [comment]
Hi Ivan, I was wondering - does the issue persist? did you find the cause?

**Comment by ivangusev** [comment]
Hi Bo,
sorry for no reply from my side. just busy these days.

It seems for me the test you did does not have logic i explained in issue:
i take A.pdf, combine it with B.pdf, save result to A.pdf, memory usage 30Mb, i take M.pdf(with size less than A.pdf), combine it with N.pdf, save result to M.pdf, memory usage 50MB, if i continue this process then all free memory will be taken.

If possible please implement this logic.
May be on weekend i will have time to try myself.

If you interested what solution i use for my script, here it is:
`pdftk doc1.pdf doc2.pdf cat output doc-combine-1-2-pdftk.pdf`
Is working well, but slower than your gem.
That's why i would like to replace it with combine_pdf.
You asked about version: combine_pdf-0.2.16

**Comment by boazsegev** [comment]
I notice that when you are using `pdftk`, you are saving the output to a file with a different name (`doc-combine-1-2-pdftk.pdf `) and, in contrast, when you're using `CombinePDF`, you're saving the output file to a file with the same name as the input file. I think maybe this is effecting the issue.

What happens if you use `CombinePDF` the same way you use `pdftk`?

What happened if you use `pdftk` the same way you use `CombinePDF` (use the same file name for the output)?

P.S.

I'm not sure I understand how to implement the logic you described without having my PDF files growing all the time... Let me know what you try and the results, because I can't reproduce the issue (I'm also using version 0.2.16).

**Comment by ivangusev** [comment]
Hi Bo,

Problem of "stupid" pdftk is that not possible to save file with same name. 
So on next line i rename doc-combine-1-2-pdftk.pdf to 1.pdf (is extra operation to do).
This is another reason i would like to move to comibnePDF.
But i use pdftk absolutely same way.

**Comment by ivangusev** [comment]
Hi Bo,

Problem of "stupid" pdftk is that not possible to save file with same name. 
So on next line i rename doc-combine-1-2-pdftk.pdf to 1.pdf (is extra operation to do).
This is another reason i would like to move to comibnePDF.
But i use pdftk absolutely same way.

**Comment by ivangusev** [comment]
Hi Bo,

I have done several tests with alone combinePDF code. With irb and daemon. And no issues happened.
So only idea i have is that there is something related to big pdf files i have on production server. This requires additional testing.

**Comment by boazsegev** [comment]
I'm very thankful for the update and I'm leaving the issue open for now, in case you get more data.

**Comment by boazsegev** [comment]
The issue has gone stale. If you still experience issues with memory leaks, please feel free to reopen the issue or post any new data you might have, so I can work on a solution.

**Comment by IonutPerfectWeb** [comment]
I have the same problem, the code I am using is : 
NOTE: pages = [1, 2, 4, 5, 6, 8] 
```
 document_name = doc_name
    existing_pdf = CombinePDF.parse Net::HTTP.get_response(URI.parse(document)).body
    new_pdf = CombinePDF.new
    pages.each do |page|
      new_pdf << existing_pdf.page(page)
    end
    new_pdf.save(document_name)
```

the pdf i am using is this one: http://mirror.hmc.edu/ctan/macros/latex/contrib/lipsum/lipsum.pdf

I am loading them from an external source, and receiving this warning when trying to parse: 
PDF 1.5 Object streams found - they are not fully supported! attempting to extract objects.


Then, while trying to parse it, all my 16Gb of rams are getting filled up, and the pc hangs...

**Comment by boazsegev** [comment]
I think this might be a Ruby interpreter issue...

The code in the loop throws an exception (there's no method "page", there's one called "pages") but the loop never stops.

It works as expected when using the correct code (notice the square brackets too):

      new_pdf << existing_pdf.pages[page]

**Comment by IonutPerfectWeb** [comment]
Yes yes, I realized my mistake and fixed it, that's why i removed the comment. Thank you for the reply

**Comment by boazsegev** [comment]
Sorry, I replied from email and didn't notice. I see that now...

**Comment by Rim-777** [comment]
Hi Bo,
Could you please clarify is it possible to save combined file into memory as Tempfile(Ruby)?
In my case I have to combine pdf files(received from an another server)  to single file and store this file into CDN  - service.

**Comment by boazsegev** [comment]
Hi @Rim-777 ,

I'll start by pointing out that I don't think this is the best place for the question.

By placing the question here, all the people that ever commented on the issue will get a copy of our discussion (which probably just clutters their emails for no reason).  You can either open a new issue for your question or ask it on Stack Overflow...

Also, I'm not sure I understand your question.

Do you mean "Can I save the PDF data to a Tempfile?" or do you mean "Can I use a Tempfile instead of in-memory objects?"

CombinePDF can be rendered as a binary PDF data (String object), which can be saved to a Tempfile or sent as a memory buffer.

However, if you need to minimize the memory footprint and use a Tempfile (instead of memory) while  CombinePDF is parsing and processing, than the answer is that CombinePDF doesn't support that option.

I hope this help. If you have further questions, please feel free to ask using a new issue or on Stack Overflow.

Kindly,
  B.
